### PR TITLE
Fix RA fix

### PR DESCRIPTION
**Pull request by ThomasHaas**
This PR fixes a bug in the fix from #523.
The fix is still ugly but now it should be correct at least :).

### CODE REVIEW

**Comment by ThomasHaas** [comment]
> The removal in the `toAdd` delta seems to explicitly handle the two bullet points of the "Impact" section in #523.
> 
> Is this what that issue refers as "propagate decremental updates"? If so, didn't you say that solution would not work for recursive relations?

No, decremental updates are if you propagate a deletion, i.e., you try to reduce already computed results.
What this "fix" does is to reduce the incremental update before applying it so that we never generate too large results in the first place.

The problematic situation is with a relation `let c = a \ b`.
What happens right now is that when `a` gets incremented by `delta(a)`, the propagator will sent the update `delta(c) = delta(a) \ b`. This is fine if `b` was already fully computed, but that may not necessarily be the case, e.g., `b` might still be empty. If now `b` gets incremented by `delta(b)`, we would need to propagate a decremental update to `c` to delete some elements, but that does not work in general (so we don't do it).
However, the updates are processed along a stratification meaning that when we merge `delta(c)` into `c` (i.e., perform the update), we actually know that `b` is fully computed (cause `b` must be in a smaller stratum since we don't allow differences in recursion).
So we can fix the proper update to `delta(c) <- delta(c) \ b` before applying it.
In short: an update to `a` may propagate a too large update to `c` but we can later on correct this update when we actually merge it into `c`.

**Comment by hernanponcedeleon** [comment]
"We don't allow difference in recursion" you mean "The right hand side of a difference cannot be recursive"?

Does this means that this solution should be fully sound as long as we don't allow those?

What about the comment saying that the push solution does not properly respect the stratification? Does it refers to this or something different?

**Comment by ThomasHaas** [comment]
> "We don't allow difference in recursion" you mean "The right hand side of a difference cannot be recursive"?

Yes.

> Does this means that this solution should be fully sound as long as we don't allow those?

I believe so.

> What about the comment saying that the push solution does not properly respect the stratification? Does it refers to this or something different?

Our current "push-based" approach to propagation does not respect the stratification properly in the sense that it can push updates from a lower stratum to a higher stratum before all "intermediate" strata were processed (i.e., any stratum that is inbetween the lower and the higher one). This does not matter if there are no differences/negations involved because then any chaotic iteration strategy will give you the correct result.
A "pull-based" approach would not have this problem with differences because an update for the higher stratum gets computed by pulling data from all lower strata, which are all fully computed by that point in time.

The fix essentially makes sure that at least for a difference, we locally also "pull" changes. What I mean by this is that when `c` (in `let c = a \ b`) gets processed, we take all the updates that were pushed onto `c` (from `a`) and reduce them by pulling data from `b`.
An alternative fix would be to make sure that `b` is always before `a` in the stratification (I haven't tested this approach).

So the comment is meant to say 
```
Our whole propagation strategy is push-based and thus fails to respect stratification properly.
This does not matter for positive relations (stratification is irrelevant) but can cause wrong results on negative relations (i.e., differences), hence we introduce this "ugly" fix.
A pull-based propagation approach would not have needed this fix in the first place.
```

The more I think about this though, the less I'm sure about the "pull-based" approach being better in general.
It clearly wins for non-recursive relations, but it feels more complicated to do on recursive one's.
In fact, when I implemented propagation in the CAAT solver, I ended up having a hybrid approach that kinda does both. It actually handles differences pretty much the same way this fix does.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java [hunk 1]
```diff
@@ -409,6 +422,19 @@ public static final class Delta {
             may = maySet;
             must = mustSet;
         }
+
+        private static Delta combine(List<Delta> deltas) {
+            if (deltas.size() == 1) {
+                return deltas.get(0);
+            }
+            EventGraph mayDelta = new EventGraph();
+            EventGraph mustDelta = new EventGraph();
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java:431]: It will always create a new object, even if must sets are disabled and always point to the signleton. Probably, will not have any visible impact in practice.

**Review by natgavrilenko — APPROVED** [review]

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java (hunk 1 above)
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java:431]: Yes, to streamline the handling of the `enableMustSets` option, I changed the code to just always compute the must sets and then just filter them out in the core algorithm before propagating.
This way the option is handled at a central place using only a few lines of code.
I believe the performance impact is marginal.
